File: src/services/executor.js
const { spawn } = require('child_process');
const { addJobLog, updateJobStatus, getJobById } = require('./jobs');
const { loadPipelineDefinition } = require('./pipeline');

function writeCommandOutput(id, level, chunk) {
  const text = chunk.toString().trim();

  if (!text) {
    return;
  }

  for (const line of text.split(/\r?\n/)) {
    const message = line.trim();

    if (!message) {
      continue;
    }

    addJobLog(id, {
      level,
      message,
    });
  }
}

function runCommand(id, command, cwd) {
  return new Promise((resolve, reject) => {
    const child = spawn(command, {
      cwd,
      env: process.env,
      shell: true,
    });

    child.stdout.on('data', (chunk) => {
      writeCommandOutput(id, 'info', chunk);
    });

    child.stderr.on('data', (chunk) => {
      writeCommandOutput(id, 'error', chunk);
    });

    child.on('error', (error) => {
      reject(error);
    });

    child.on('close', (code, signal) => {
      resolve({
        code,
        signal,
      });
    });
  });
}

async function runJob(id) {
  const runningResult = updateJobStatus(id, 'running');

  if (!runningResult.ok) {
    return runningResult;
  }

  addJobLog(id, {
    level: 'info',
    message: 'Executor picked up job from queue',
  });

  const currentJob = getJobById(id);
  const workspacePath = currentJob?.workspacePath || process.cwd();
  const pipelineFile = currentJob?.pipelineFile || '.relay.yml';

  addJobLog(id, {
    level: 'info',
    message: `Using workspace ${workspacePath}`,
  });

  const pipelineResult = loadPipelineDefinition(workspacePath, pipelineFile);

  if (!pipelineResult.ok) {
    addJobLog(id, {
      level: 'error',
      message: `Failed to load pipeline: ${pipelineResult.reason}`,
    });

    return updateJobStatus(id, 'failed');
  }

  addJobLog(id, {
    level: 'info',
      message: `Loaded pipeline from ${pipelineResult.pipelinePath}`,
    });

  for (const step of pipelineResult.pipeline.steps) {
    addJobLog(id, {
      level: 'info',
      message: `Starting step "${step.name}"`,
    });

    addJobLog(id, {
      level: 'info',
      message: `Command: ${step.command}`,
    });

    let result;

    try {
      result = await runCommand(id, step.command, workspacePath);
    } catch (error) {
      addJobLog(id, {
        level: 'error',
        message: `Step "${step.name}" crashed before completion: ${error.message}`,
      });

      return updateJobStatus(id, 'failed');
    }

    if (result.code !== 0) {
      addJobLog(id, {
        level: 'error',
        message: `Step "${step.name}" failed with exit code ${result.code}${
          result.signal ? ` and signal ${result.signal}` : ''
        }`,
      });

      return updateJobStatus(id, 'failed');
    }

    addJobLog(id, {
      level: 'info',
      message: `Step "${step.name}" completed successfully`,
    });
  }

  addJobLog(id, {
    level: 'info',
    message: 'Executor finished all pipeline steps',
  });

  return updateJobStatus(id, 'passed');
}

module.exports = {
  runJob,
};

File: app/services/executor.py
from __future__ import annotations

from subprocess import Popen, PIPE
from threading import Thread
import random
import time

from .jobs import add_job_log, get_job_by_id, update_job_status
from .pipeline import load_pipeline_definition


def _stream_output(job_id: str, pipe, level: str) -> None:
    for line in iter(pipe.readline, ""):
        message = line.strip()
        if message:
            add_job_log(job_id, level=level, message=message)
    pipe.close()


def run_job(job_id: str, worker: dict | None = None) -> dict:
    running_result = update_job_status(job_id, "processing")
    if not running_result["ok"]:
        return running_result

    if worker:
        add_job_log(
            job_id,
            level="info",
            message=f'Worker {worker["name"]} picked up job for {worker["supportedLanguages"]}',
        )
        startup_delay = round(random.uniform(0.15, 0.75) * worker.get("speedMultiplier", 1.0), 2)
        add_job_log(job_id, level="info", message=f"Worker startup delay: {startup_delay}s")
        time.sleep(startup_delay)
    else:
        add_job_log(job_id, level="info", message="Executor picked up job from queue")

    job = get_job_by_id(job_id)
    workspace_path = job["workspacePath"]
    pipeline_file = job["pipelineFile"]

    add_job_log(job_id, level="info", message=f"Using workspace {workspace_path}")

    pipeline_result = load_pipeline_definition(workspace_path, pipeline_file)
    if not pipeline_result["ok"]:
        add_job_log(job_id, level="error", message=f'Failed to load pipeline: {pipeline_result["reason"]}')
        return update_job_status(job_id, "failed")

    add_job_log(job_id, level="info", message=f'Loaded pipeline from {pipeline_result["pipeline_path"]}')

    for step in pipeline_result["pipeline"]["steps"]:
        add_job_log(job_id, level="info", message=f'Starting step "{step["name"]}"')
        add_job_log(job_id, level="info", message=f'Command: {step["command"]}')

        if worker:
            jitter = round(random.uniform(0.05, 0.4) * worker.get("speedMultiplier", 1.0), 2)
            add_job_log(job_id, level="info", message=f'Worker {worker["name"]} warmup jitter: {jitter}s')
            time.sleep(jitter)

        try:
            process = Popen(
                step["command"],
                cwd=workspace_path,
                shell=True,
                text=True,
                stdout=PIPE,
                stderr=PIPE,
            )
        except Exception as error:  # noqa: BLE001
            add_job_log(job_id, level="error", message=f'Step "{step["name"]}" crashed before completion: {error}')
            return update_job_status(job_id, "failed")

        stdout_thread = Thread(target=_stream_output, args=(job_id, process.stdout, "info"), daemon=True)
        stderr_thread = Thread(target=_stream_output, args=(job_id, process.stderr, "error"), daemon=True)
        stdout_thread.start()
        stderr_thread.start()
        code = process.wait()
        stdout_thread.join()
        stderr_thread.join()

        if code != 0:
            add_job_log(job_id, level="error", message=f'Step "{step["name"]}" failed with exit code {code}')
            return update_job_status(job_id, "failed")

        add_job_log(job_id, level="info", message=f'Step "{step["name"]}" completed successfully')

    add_job_log(job_id, level="info", message="All pipeline steps completed")
    return update_job_status(job_id, "processed")
